### PR TITLE
[brave-today] Feed item parsing clean up

### PR DESCRIPTION
**Pull request by cdesouza-chromium**
This PR introduces an IDL file to handle `base::Value` checks for feed parsing. The current code has a number of dereferences in the code that pose a significant risk, and to guard against them will involve even a higher number of awkward control flow, which can be quite easy to get wrong.

Another PR will follow this one converting other parsts of the feed to also use a type generated by the json-schema-compiler for parsing/validation.

Resolves https://github.com/brave/brave-browser/issues/28654

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_news/browser/combined_feed_parsing.cc
Comment (petemill) [inline components/brave_news/browser/combined_feed_parsing.cc:1]: It would be a shame to lose the git history link to the previous file. Is it possible to split in to two commits - 1 that moves the file, and one that modifies?

File: components/brave_news/api/feed.idl
Comment (petemill) [inline components/brave_news/api/feed.idl:1]: Since this is only for the brave-combined feed, and not either the internal feed representation (as defined in mojom) or the remote direct feed items, how about we name this to match the new combined_feed_parsing.cc file, so perhaps `combined_feed`?

File: components/brave_news/browser/combined_feed_parsing.cc
Comment (cdesouza-chromium) [reply components/brave_news/browser/combined_feed_parsing.cc:1]: I have submitted it as a separate PR. I will rebase this one on top once that one is merged. https://github.com/brave/brave-core/pull/17424

File: components/brave_news/api/feed.idl
Comment (cdesouza-chromium) [reply components/brave_news/api/feed.idl]: Done.

**Comment by bsclifton** [comment]
@cdesouza-chromium updated the labels on the issue (QA/no, release notes/excluse) and ticked off the items in checklist above 😄 Let me know if those look good

**Comment by cdesouza-chromium** [comment]
@bsclifton I think my user rights to set labels on issues has been removed. The field is not available to me anymore on the issues I'm opening.

**Review by petemill — APPROVED** [review]
Thanks for taking this on! Looks great. Just 1 comment about `NOTREACHED` and its impact which I'll let you decide.

**Pull request merged**
